sdk/core/core/src/az_json_read.c

```diff
@@ -0,0 +1,379 @@
+      return AZ_JSON_ERROR_UNEXPECTED_CHAR;
+    }
+  }
+  return AZ_OK;
+}
+
+az_result az_json_read_number_digit_rest(
+  az_const_str const buffer,
+  size_t *const p,
+  double *const out_value,
+  int const sign
+) {
+  double value = 0;
+
+  // integer part
+  {
+    char c = az_const_str_item(buffer, *p);
+    if (c != '0') {
+      // read an integer part of the number
+      do {
+        value = value * 10 + (c - '0');
+        *p += 1;
+        if (*p == buffer.size) {
+          break;
+        }
+        c = az_const_str_item(buffer, *p);
+      } while (az_is_digit(c));
+    } else {
+      *p += 1;
+    }
+  }
+
+  // fraction
+  int16_t exp = 0;
+  if (*p != buffer.size && az_const_str_item(buffer, *p) == '.') {
+    *p += 1;
+    if (*p == buffer.size) {
+      return AZ_JSON_ERROR_UNEXPECTED_END;
+    }
+    char c = az_const_str_item(buffer, *p);
+    if (!az_is_digit(c)) {
+      return AZ_JSON_ERROR_UNEXPECTED_CHAR;
+    }
+    do {
+      value = value * 10 + (c - '0');
+      --exp;
+      *p += 1;
+      if (*p == buffer.size) {
+        break;
+      }
+      c = az_const_str_item(buffer, *p);
+    } while (az_is_digit(c));
+  }
+
+  // exp
+  if (*p != buffer.size) {
+    switch (az_const_str_item(buffer, *p)) {
+      case 'e':
+      case 'E':
+        {
+          // skip 'e' or 'E'
+          *p += 1;
+
+          if (*p == buffer.size) {
+            return AZ_JSON_ERROR_UNEXPECTED_END;
+          }
+          char c = az_const_str_item(buffer, *p);
+
+          // read sign, if any.
+          int e_sign = 1;
+          switch (c) {
+            case '-':
+              e_sign = -1;
+            case '+':
+              *p += 1;
+              if (*p == buffer.size) {
+                return AZ_JSON_ERROR_UNEXPECTED_END;
+              }
+              c = az_const_str_item(buffer, *p);
+          }
+
+          // expect at least one digit.
+          if (!az_is_digit(c)) {
+            return AZ_JSON_ERROR_UNEXPECTED_CHAR;
+          }
+
+          int16_t e_int = 0;
+          do {
+            e_int = e_int * 10 + (c - '0');
+            *p += 1;
+            if (*p == buffer.size) {
+              break;
+            }
+            c = az_const_str_item(buffer, *p);
+          } while (az_is_digit(c));
+          exp += e_int * e_sign;
+        }
+    }
+  }
+
+  *out_value = value * pow(10.0, exp) * sign;
+  return AZ_OK;
+}
+
+az_result az_json_read_number_minus_rest(az_const_str const buffer, size_t *const p, double *const out_value) {
+  // skip '-'
+  *p += 1;
+  // read digit.
+  char c;
+  AZ_RETURN_ON_ERROR(az_json_get_char(buffer, *p, &c));
+  return az_is_digit(c) ? az_json_read_number_digit_rest(buffer, p, out_value, -1) : AZ_JSON_ERROR_UNEXPECTED_CHAR;
+}
+
+az_result az_json_read_string_rest(az_const_str const buffer, size_t *const p, az_json_string *const string) {
+  // skip '"'
+  size_t const begin = *p;
+  string->begin = buffer.begin + begin;
```

Comment:
this is very wierd since az_json_string is a const string. It's allowed here because begin is `const char*` and not `const char * const` but still....

Reply:
eagh I can live with it I guess.